File: little-bee/dao/config/alarm.go
package config

import (
	"dao/database"
	"message/littlebee"
	"message/querycommand"
	"util/queryclause"

	"github.com/fatih/structs"
)

func CreateAlarm(l *littlebee.Alarm) (*littlebee.Alarm, error) {
	result := database.DB.Create(l)
	return l, result.Error
}

func RetrieveAlarm(alarmId int32) (*littlebee.Alarm, error) {
	r := &littlebee.Alarm{}
	result := database.DB.First(&r, alarmId)
	return r, result.Error
}

func EndAlarmBySerialNumber(l *littlebee.Alarm) error {
	result := database.DB.Model(l).Where("alarm_serial_number = ?", l.AlarmSerialNumber).Select("AlarmEndTime", "AlarmDesc").Updates(l)
	return result.Error
}

func EndAlarmByAlarmTime(l *littlebee.Alarm) error {
	result := database.DB.Model(l).Where("signal_id = ? AND alarm_begin_time = ?", l.AlarmBeginTime, l.AlarmSerialNumber).Select("AlarmEndTime", "AlarmDesc").Updates(l)
	return result.Error
}

func UpdateAlarm(l *littlebee.Alarm) error {
	lMap := structs.Map(l)
	result := database.DB.Model(l).Omit("AlarmId", "CreatedAt", "CreatedBy", "IsDeleted", "DeletedAt", "DeletedBy").Updates(lMap)
	return result.Error
}

func AckAlarm(l *littlebee.Alarm) error {
	lMap := structs.Map(l)
	lMap["alarm_begin_time"] = l.AlarmBeginTime.AsTime().UTC()
	lMap["ack_time"] = l.AckTime.AsTime().UTC()
	result := database.DB.Model(l).Where("signal_id = ? AND alarm_begin_time = ? AND ack_state = 0", l.SignalId, lMap["alarm_begin_time"]).Select("AckState", "AckTime", "AckUserId", "AckUserName").Updates(lMap)
	return result.Error
}

func DeleteAlarmWithFlag(l *littlebee.Alarm) error {
	result := database.DB.Model(l).Updates(littlebee.Alarm{})
	return result.Error
}

func DeleteAlarm(l *littlebee.Alarm) error {
	result := database.DB.Delete(l)
	return result.Error
}

func QueryAlarm(q *querycommand.QueryCommand) (*littlebee.AlarmList, error) {
	var alarms []*littlebee.Alarm
	s := "SELECT * FROM alarms "
	w, v := queryclause.QueryCommandString(q)
	s = s + w
	result := database.DB.Raw(s, v...).Scan(&alarms)
	return &littlebee.AlarmList{Items: alarms}, result.Error
}

func GetAllAlarms() (*littlebee.AlarmList, error) {
	var alarms []*littlebee.Alarm
	result := database.DB.Find(&alarms)
	return &littlebee.AlarmList{Items: alarms}, result.Error
}

func EndAlarm(l *littlebee.Alarm) error {
	lMap := structs.Map(l)
	result := database.DB.Model(l).Omit("AlarmId", "CreatedAt", "CreatedBy", "IsDeleted", "DeletedAt", "DeletedBy").Updates(lMap)
	return result.Error
}

File: frontend/console/src/api/alarm.js
import Axios from "@/http";

export function createAlarm(params) {
    return new Promise((resolve, reject) => {
        Axios.post('/config/alarm', params, { isJSON: true }).then(res => {
            resolve(res)
        })
    })
}

export function retrieveAlarm(alarmId) {
    return new Promise((resolve, reject) => {
        Axios.get(`/config/alarm/${ alarmId }`).then(res => {
            resolve(res)
        })
    })
}

export function updateAlarm(params) {
    return new Promise((resolve, reject) => {
        Axios.put('/config/alarm', params, { isJSON: true }).then(res => {
            resolve(res)
        })
    })
}

export function deleteAlarm(alarmId) {
    return new Promise((resolve, reject) => {
        Axios.delete(`/config/alarm/${ alarmId }`).then(res => {
            resolve(res)
        })
    })
}

export function deleteAlarmWithFlag(alarmId) {
    return new Promise((resolve, reject) => {
        Axios.put(`/config/alarm/${ alarmId }`).then(res => {
            resolve(res)
        })
    })
}

export function getAllAlarms() {
    return new Promise((resolve, reject) => {
        Axios.get('/config/alarm/all').then(res => {
            resolve(res)
        })
    })
}

export function queryAlarm (params) {
    return new Promise((resolve, reject) => {
        Axios.post('/config/alarm/query', params, { isJSON: true }).then(res => {
            resolve(res)
        })
    })
}

export function ackAlarm(params) {
    return new Promise((resolve, reject) => {
        Axios.post('/config/alarm/ack', params, { isJSON: true }).then(res => {
            resolve(res)
        })
    })
}